actors/migration/top.go

```diff
@@ -17,65 +23,159 @@ type StateMigration interface {
 		StateMigration: &marketMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.StorageMinerActorCodeID,
+	builtin0.StorageMinerActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.StorageMinerActorCodeID,
 		StateMigration: &minerMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.MultisigActorCodeID,
+	builtin0.MultisigActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.MultisigActorCodeID,
 		StateMigration: &multisigMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.PaymentChannelActorCodeID,
+	builtin0.PaymentChannelActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.PaymentChannelActorCodeID,
 		StateMigration: &paychMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.StoragePowerActorCodeID,
+	builtin0.StoragePowerActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.StoragePowerActorCodeID,
 		StateMigration: &powerMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.RewardActorCodeID,
+	builtin0.RewardActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.RewardActorCodeID,
 		StateMigration: &rewardMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.SystemActorCodeID,
+	builtin0.SystemActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.SystemActorCodeID,
 		StateMigration: &systemMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.VerifiedRegistryActorCodeID,
+	builtin0.VerifiedRegistryActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.VerifiedRegistryActorCodeID,
 		StateMigration: &verifregMigrator{},
 	},
 }
+
+// A phoenix tracks the unburning of funds
+type phoenix struct {
+	burntBalance abi.TokenAmount
+}
+
+func (p phoenix) load(ctx context.Context, actorsIn *states.TreeTop) error {
+	burntFundsActor, err := actorsIn.GetActor(ctx, builtin0.BurntFundsActorAddr)
+	if err != nil {
+		return err
+	}
+	p.burntBalance = burntFundsActor.Balance
+	return nil
+}
+
+func (p phoenix) curr() abi.TokenAmount {
+	return p.burntBalance
+}
+
+func (p phoenix) transfer(amt abi.TokenAmount) error {
+	p.burntBalance = big.Sub(p.burntBalance, amt)
+	if p.burntBalance.LessThan(big.Zero()) {
+		return xerrors.Errorf("migration programmer error burnt funds balance falls to %v, below zero", p.burntBalance)
+	}
+	return nil
+}
+
+func (p phoenix) flush(ctx context.Context, actorsIn, actorsOut *states.TreeTop) error {
+	burntFundsActor, err := actorsIn.GetActor(ctx, builtin0.BurntFundsActorAddr)
+	if err != nil {
+		return err
+	}
+	burntFundsActor.Balance = p.burntBalance
+	return actorsOut.SetActor(ctx, builtin.BurntFundsActorAddr, burntFundsActor)
+}
+
+// Migrates the filecoin state tree starting from the global state tree and upgrading all actor state.
+func MigrateStateTree(ctx context.Context, store cbor.IpldStore, stateRootIn cid.Cid) (cid.Cid, error) {
+	// first migrate the global state tree hamt to something v2 can work with
+	// if this is very slow (likely) we can cherry-pick states.TreeTop to v0.9
+	// to avoid this step.
+	stateRootInTweaked, err := migrateHAMTRaw(ctx, store, stateRootIn)
```

Comment:
I don't understand why you're doing this first. We're going to rewrite the whole thing again anyway. Why re-write it to something that needs to be rewritten yet again. This will be slow and pollute the store.

Reply:
The reason is that the new hamt dependency will possibly be unable to read the old HAMT because with high probability there are keys out of order.  I think you've realized this with the all to pull out the state tree changes on v0.9.  As the comment states, if we do that we can avoid this migration.

Reply:
Ok, #1147